apps/openchallenges/app/src/app/custom-scroll-restoration.service.ts

```diff
@@ -0,0 +1,71 @@
+import { ViewportScroller } from '@angular/common';
+import { Injectable } from '@angular/core';
+import { Event, Router, RoutesRecognized, Scroll } from '@angular/router';
+import { combineLatest, filter, pairwise } from 'rxjs';
+
+/**
+ * A service for restoring scroll positions based on the method of navigation.
+ * It helps preserve the user's scroll position during various types of navigation actions.
+ * - Backward navigation: Scrolls to the previous position.
+ * - Anchor navigation: Scrolls to a specified anchor.
+ * - Param navigation: Maintains the current scroll position,
+ * - Forward navigation/Same url navigation: Scrolls to the top of the page.
+ */
+
+@Injectable({
+  providedIn: 'root',
+})
+export class CustomScrollRestorationService {
+  private scrollX = 0;
+  private scrollY = 0;
+
+  constructor(
+    private router: Router,
+    private viewportScroller: ViewportScroller
+  ) {}
+
+  scrollPositionRestoration(): void {
+    // Capture the current scroll position
+    window.addEventListener('scroll', () => {
```

Comment:
Is there a way to do this without using `window`? `window` is not available when rendering pages on the server (SSR). Hence, this code will fail is the method is called during server-side rendering. I will give it a try and report my feedback in this PR.

Reply:
Not sure if it will help, but I have changed to get current scroll position directly via `viewportScroller`, instead of `window`.